### PR TITLE
Feature/intersection tables

### PR DESCRIPTION
**Pull request by jacob6838**
# PR Details

## Description

**DISCLAIMER** This is not intended to be included in the current JPO-USDOT release. As such, this has been created as a Draft PR.

This PR adds the management of intersection data to the cvmanager. This includes the following:
1. Adding postgresql tables for intersections, intersection_organizations, and rsu_intersections
2. Adding API endpoints to handle management of intersections and links
3. Adding webapp Admin pages for creating, editing, and removing intersections, intersection-organization links, and rsu-intersection links.

## How Has This Been Tested?

These changes were tested locally, running the api, webapp, and database updates together. Through the UI, you are able to create intersections, remove intersections, edit intersections, create/remove intersection - organization links, and create/remove rsu - intersection links. 

## Types of changes

- [ ] Defect fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that cause existing functionality to change)

## Checklist:

- [ ] My changes require new environment variables:
  - [ ] I have updated the docker-compose, K8s YAML, and all dependent deployment configuration files.
- [ ] My changes require updates to the documentation:
  - [ ] I have updated the documentation accordingly.
- [x] My changes require updates and/or additions to the unit tests:
  - [x] I have modified/added tests to cover my changes.
- [x] All existing tests pass.

### CODE REVIEW

File: resources/sql_scripts/CVManager_CreateTables.sql [hunk 1]
```diff
@@ -308,6 +308,7 @@ CREATE TABLE IF NOT EXISTS public.rsu_organization
       ON DELETE NO ACTION
 );
 
+-- TODO: should combine with intersection table?
```
Comment (dmccoystephenson) [inline resources/sql_scripts/CVManager_CreateTables.sql]: Should this TODO be taken care of in this PR? If not, should we open a work item for this?

**Review by dmccoystephenson — CHANGES_REQUESTED** [review]
The API unit tests have some failures:

image

File: resources/sql_scripts/CVManager_CreateTables.sql (hunk 1 above)
Comment (jacob6838) [reply resources/sql_scripts/CVManager_CreateTables.sql]: Replaced this with a new work item: https://dev.azure.com/SOC-OIT/CDOT/_workitems/edit/263317

**Review by dmccoystephenson — APPROVED** [review]
This looks good to me! The API/webapp unit tests pass & I am able to access the new UI pages through the webapp.

File: resources/sql_scripts/CVManager_CreateTables.sql [hunk 2]
```diff
@@ -459,4 +459,70 @@ CREATE TABLE IF NOT EXISTS public.obu_ota_requests (
 	CONSTRAINT fk_manufacturer FOREIGN KEY (manufacturer) REFERENCES public.manufacturers(manufacturer_id)
 );
 
-CREATE SCHEMA IF NOT EXISTS keycloak;
\ No newline at end of file
+CREATE SCHEMA IF NOT EXISTS keycloak;
```
Comment (drewjj) [inline resources/sql_scripts/CVManager_CreateTables.sql:462]: Can you add a SQL script with just these new table CREATEs in the update_scripts directory? Will be handy for making the updates without much thought.

File: services/api/src/admin_org.py [hunk 3]
```diff
@@ -270,6 +329,20 @@ def check_orphan_rsus(org):
     return False
 
 
+def check_orphan_intersections(org):
```
Comment (drewjj) [inline services/api/src/admin_org.py:332]: Nice, glad this is being accounted for.

File: webapp/src/features/adminEditIntersection/AdminEditIntersection.tsx [hunk 4]
```diff
@@ -0,0 +1,334 @@
...
+  const navigate = useNavigate()
+
+  const {
+    register,
+    handleSubmit,
+    formState: { errors },
+    setValue,
+  } = useForm<AdminEditIntersectionFormType>({
+    defaultValues: {
+      intersection_id: '',
```
Comment (drewjj) [inline webapp/src/features/adminEditIntersection/AdminEditIntersection.tsx:61]: Is there much value to allowing intersection ID to be seen on the modification when the API doesn't actually allow for this to be modified?

File: webapp/src/features/adminAddIntersection/AdminAddIntersection.tsx [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+    <Dialog open={open}>
+      <DialogTitle>Add Intersection</DialogTitle>
+      <DialogContent>
+        <Form
+          id="add-intersection-form"
+          onSubmit={handleSubmit((data) => handleFormSubmit(data))}
+          style={{ fontFamily: 'Arial, Helvetica, sans-serif' }}
+        >
+          <Form.Group className="mb-3" controlId="ip">
+            <Form.Label>Intersection ID</Form.Label>
```
Comment (drewjj) [inline webapp/src/features/adminAddIntersection/AdminAddIntersection.tsx:92]: Is the intersection ID necessary to allow the user to define it? Is this the value that gets used to match corresponding MAP and SPaT messages to?

File: services/api/tests/src/test_admin_org.py [hunk 6]
```diff
@@ -282,20 +286,52 @@ def test_delete_org(mock_query_db, mock_write_db):
     mock_write_db.assert_has_calls(calls)
     assert actual_result == expected_result
 
+
 @patch("api.src.admin_org.pgquery.query_db")
 def test_delete_org_failure_orphan_rsu(mock_query_db):
```
Comment (drewjj) [inline services/api/tests/src/test_admin_org.py:291]: While the API unit tests pass, we seem to be missing tests for all of the new python scripts you've added. We will want to  add some tests for these.

**Review by drewjj — CHANGES_REQUESTED** [review]
Everything was put together nicely and the unit tests pass for both the webapp and api. A number of comments I would like addressed but it is very close!

File: webapp/src/features/adminEditIntersection/AdminEditIntersection.tsx (hunk 4 above)
Comment (jacob6838) [reply webapp/src/features/adminEditIntersection/AdminEditIntersection.tsx:61]: The intersection_id corresponds to the intersection_number in the sql table. The intersection_id (intersection_number in sql) I believe should be able to be changed

Comment (jacob6838) [reply webapp/src/features/adminEditIntersection/AdminEditIntersection.tsx:61]: Added the ability to modify the intersection ID

File: webapp/src/features/adminAddIntersection/AdminAddIntersection.tsx (hunk 5 above)
Comment (jacob6838) [reply webapp/src/features/adminAddIntersection/AdminAddIntersection.tsx:92]: This is the intersection_number in the database, used to link up MAP and SPAT messages

File: resources/sql_scripts/CVManager_CreateTables.sql (hunk 2 above)
Comment (jacob6838) [reply resources/sql_scripts/CVManager_CreateTables.sql:462]: added!

File: services/api/tests/src/test_admin_org.py (hunk 6 above)
Comment (jacob6838) [reply services/api/tests/src/test_admin_org.py:291]: Added and passing

**Review by drewjj — APPROVED** [review]
Nice, everything looks great!

File: services/api/tests/src/test_admin_org.py [hunk 7]
```diff
@@ -217,6 +219,8 @@ def test_modify_user_success(mock_pgquery, mock_check_safe_input):
         call(admin_org_data.modify_org_remove_user_sql),
         call(admin_org_data.modify_org_add_rsu_sql),
         call(admin_org_data.modify_org_remove_rsu_sql),
+        call(admin_org_data.modify_org_add_intersection_sql),
```
Comment (payneBrandon) [inline services/api/tests/src/test_admin_org.py:222]: Just noticing with these additions, but we're calling admin_org.modify_org and the test method is called "test_modify_user_success". Should we rename to reflect what we're really doing here?

File: services/api/tests/src/test_admin_new_intersection.py [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+        status = admin_new_intersection.AdminNewIntersection()
+        (body, code, headers) = status.post()
+
+        mock_add_intersection.assert_called_once()
+        assert code == 200
+        assert headers["Access-Control-Allow-Origin"] == "test.com"
+        assert body == {}
+
+
+def test_entry_post_schema():
```
Comment (payneBrandon) [inline services/api/tests/src/test_admin_new_intersection.py]: Are we testing a bad schema here? Should we name it to reflect?

File: services/api/tests/src/test_admin_intersection.py [hunk 9]
```diff
@@ -0,0 +1,285 @@
+from unittest.mock import patch, MagicMock, call
```
Comment (payneBrandon) [inline services/api/tests/src/test_admin_intersection.py:1]: This file seems to be predominantly copy/paste from test_admin_rsu.py, to the point we're even importing the admin_rsu and admin_rsu_data.

**Review by payneBrandon — CHANGES_REQUESTED** [review]
Service functionality looks good, but I have a few questions with the unit tests

File: webapp/src/features/adminAddIntersection/adminAddIntersectionSlice.tsx [hunk 10]
```diff
@@ -0,0 +1,256 @@
...
+}
+
+const initialState = {
+  apiData: {} as AdminIntersectionKeyedCreationInfo,
+  selectedOrganizations: [] as AdminIntersectionKeyedCreationInfo['organizations'],
+  selectedRsus: [] as AdminIntersectionKeyedCreationInfo['rsus'],
+  submitAttempt: false,
+}
+
+export const updateApiJson = (apiJson: AdminIntersectionCreationInfo): AdminIntersectionKeyedCreationInfo => {
```
Comment (payneBrandon) [inline webapp/src/features/adminAddIntersection/adminAddIntersectionSlice.tsx]: Can we add some descriptions to the methods? "UpdateApiJson" is fairly ambiguous as to what the method does.

File: services/api/tests/src/test_admin_org.py (hunk 7 above)
Comment (jacob6838) [reply services/api/tests/src/test_admin_org.py:222]: re-named to "test_modify_organization_success"

File: services/api/tests/src/test_admin_new_intersection.py (hunk 8 above)
Comment (jacob6838) [reply services/api/tests/src/test_admin_new_intersection.py]: renamed to "test_entry_post_schema_bad_json"

File: services/api/tests/src/test_admin_intersection.py (hunk 9 above)
Comment (jacob6838) [reply services/api/tests/src/test_admin_intersection.py:1]: Good catch, went though and updated all of these tests

File: webapp/src/features/adminAddIntersection/adminAddIntersectionSlice.tsx (hunk 10 above)
Comment (jacob6838) [reply webapp/src/features/adminAddIntersection/adminAddIntersectionSlice.tsx]: Updated the name to "convertApiJsonToKeyedFormat"

File: webapp/src/features/adminAddIntersection/adminAddIntersectionSlice.tsx [hunk 11]
```diff
@@ -0,0 +1,258 @@
...
+
+export const checkForm = (state: RootState['adminAddIntersection']) => {
+  if (state.value.selectedOrganizations.length === 0) {
+    return false
+  } else {
+    return true
+  }
+}
+
+export const updateJson = (
```
Comment (payneBrandon) [inline webapp/src/features/adminAddIntersection/adminAddIntersectionSlice.tsx]: We updated the other json conversion function name, but there's still some in here like this updateJson that we should rename and add descriptions to accordingly.

File: webapp/src/features/adminEditIntersection/adminEditIntersectionSlice.tsx [hunk 12]
```diff
@@ -0,0 +1,234 @@
...
+  rsus: [] as { name: string }[],
+  selectedRsus: [] as { name: string }[],
+  submitAttempt: false,
+}
+
+function isEmptyObject(obj: any): boolean {
+  return obj && Object.keys(obj).length === 0
+}
+
+export const checkForm = (state: RootState['adminEditIntersection']) => {
```
Comment (payneBrandon) [inline webapp/src/features/adminEditIntersection/adminEditIntersectionSlice.tsx]: Similar to previous comments, let's add some commenting into these methods to know what they are doing. "checkForm" seems to only be validating a property length is not 0, "updateJson" is removing null properties and updating some lists but without digging into the code we'd have no idea that's what it does.

File: webapp/src/features/adminEditIntersection/adminEditIntersectionSlice.tsx [hunk 13]
```diff
@@ -0,0 +1,234 @@
...
+    },
+    updateStates: (state, action: { payload: adminEditIntersectionData }) => {
+      const apiData = action.payload
+
+      const allowedSelections = apiData.allowed_selections
+      state.value.organizations = allowedSelections.organizations.map((val) => {
+        return { name: val }
+      })
+      state.value.rsus = allowedSelections.rsus.map((val) => {
+        return { name: val?.replace('/32', '') }
```
Comment (payneBrandon) [inline webapp/src/features/adminEditIntersection/adminEditIntersectionSlice.tsx]: Are we just updating the rsu ip address to be more human readable here by removing the CIDR subnet mask? Can we add some comments into these to say why we're doing some of this so we don't forget later?

File: webapp/src/features/adminAddIntersection/adminAddIntersectionSlice.tsx (hunk 11 above)
Comment (jacob6838) [reply webapp/src/features/adminAddIntersection/adminAddIntersectionSlice.tsx]: Added docstring comments and renamed method

File: webapp/src/features/adminEditIntersection/adminEditIntersectionSlice.tsx (hunk 12 above)
Comment (jacob6838) [reply webapp/src/features/adminEditIntersection/adminEditIntersectionSlice.tsx]: Added docstring comments and renamed method

File: webapp/src/features/adminEditIntersection/adminEditIntersectionSlice.tsx (hunk 13 above)
Comment (jacob6838) [reply webapp/src/features/adminEditIntersection/adminEditIntersectionSlice.tsx]: Added comments, these are both for human readability and api compatibility (for the editIntersection API)

**Comment by payneBrandon** [comment]
Overall code is looking good, but we need some more descriptions in the code, better naming conventions, or both to help the developers understand what we're looking at for methods and why things are doing what they are.

**Comment by jacob6838** [comment]
> Overall code is looking good, but we need some more descriptions in the code, better naming conventions, or both to help the developers understand what we're looking at for methods and why things are doing what they are.

I just went though and added docstring comments to all of the slice methods, and renamed the ambiguous methods

**Review by payneBrandon — APPROVED** [review]
lgtm!

**Pull request merged**